src/gatherers/critical-network-chains.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * @license
+ * Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const Gather = require('./gather');
+const log = require('../lib/log.js');
+
+const flatten = arr => arr.reduce((a, b) => a.concat(b), []);
+const includes = (arr, elm) => arr.indexOf(elm) > -1;
+
+class Node {
+  get requestId() {
+    return this.request.requestId;
+  }
+  constructor(request, parent) {
+    this.children = [];
+    this.parent = parent;
+    this.request = request;
+  }
+
+  setParent(parentNode) {
+    this.parent = parentNode;
+  }
+
+  addChild(childNode) {
+    this.children.push(childNode);
+  }
+
+  toJSON() {
+    // Prevents circular reference so we can print nodes when needed
+    return `{
+      id: ${this.requestId},
+      parent: ${!!(this.parent) && this.parent.requestId},
+      children: ${this.children.map(child => child.requestId)}
+    }`;
+  }
+
+}
+
+class CriticalNetworkChains extends Gather {
+
+  get criticalPriorities() {
+    // For now, critical request == render blocking request (as decided by
+    // blink). Blink treats requests with the following priority levels as
+    // render blocking.
+    // See https://docs.google.com/document/d/1bCDuq9H1ih9iNjgzyAL0gpwNFiEP4TZS-YLRp_RuMlc
+    return ['VeryHigh', 'High', 'Medium'];
+  }
+
+  getCriticalChains(networkRecords) {
+    // Drop the first request because it's uninteresting - it's the page html
+    // and always critical. No point including it in every request
+    const criticalRequests = networkRecords.slice(1).filter(
+      req => includes(this.criticalPriorities, req.initialPriority()));
+
+    // Build a map of requestID -> Node.
+    const requestIdToNodes = new Map();
+    for (let request of criticalRequests) {
+      const requestNode = new Node(request, null);
+      requestIdToNodes.set(requestNode.requestId, requestNode);
+    }
+
+    // Connect the parents and children
+    for (let request of criticalRequests) {
+      if (request.initiatorRequest()) {
+        const parentRequestId = request.initiatorRequest().requestId;
+        const childNode = requestIdToNodes.get(request.requestId);
+        const parentNode = requestIdToNodes.get(parentRequestId);
+        if (childNode && parentNode) {
+          // Both child and parent must be critical
```

Comment:
This is a very good question. One way this case can happen is if a low priority script ends up initiating a high priority script. I don't know the internals of blink well enough yet to answer if this case is possible. However, I really don't know if we should call the initiated script "critical" in that case, because it's very possible the page already loaded by the time we see this high priority request, and no request should be called "critical" at that point. We can improve this by checking the timestamp of requests and discarding any request that gets fired after domContentLoaded, since it's not render-blocking any more. I haven't seen this happen in practice yet though. Also, when we have better critical request data that are not just based on request priority, this issue will disappear. 

The other way this happens and is absolutely worth fixing is when there are redirects. Having a request redirected sometimes sets the priority of the new request as 'VeryLow'. For example, the first request to theverge.com has prio `VeryHigh`, but it redirects to www.theverge.com whose prio is `VeryLow`. But this new request is actually the initiator of all the subsequent requests, and we end up with a non-critical parent having critical children. I'm still trying to understand this case better and this is why I have the `TODO` about redirects.